components/places/src/db/db.rs

```diff
@@ -9,22 +9,123 @@
 use super::interrupt::{InterruptScope, PlacesInterruptHandle};
 use super::schema;
 use crate::error::*;
-use rusqlite::Connection;
+use lazy_static::lazy_static;
+use rusqlite::{Connection, OpenFlags};
 use sql_support::ConnExt;
+use std::cell::Cell;
 use std::ops::Deref;
-use std::path::Path;
+use std::path::{Path, PathBuf};
+use std::sync::Mutex;
 
 use std::sync::{atomic::AtomicUsize, Arc};
 
 pub const MAX_VARIABLE_NUMBER: usize = 999;
 
+#[derive(Debug, PartialEq)]
+pub enum ConnectionType {
+    ReadOnly,
+    ReadWrite,
+    Sync,
+}
+
+impl ConnectionType {
+    fn rusqlite_flags(&self) -> OpenFlags {
+        let common_flags = OpenFlags::SQLITE_OPEN_NO_MUTEX | OpenFlags::SQLITE_OPEN_URI;
+        match self {
+            ConnectionType::ReadOnly => common_flags | OpenFlags::SQLITE_OPEN_READ_ONLY,
+            ConnectionType::ReadWrite => {
+                common_flags | OpenFlags::SQLITE_OPEN_CREATE | OpenFlags::SQLITE_OPEN_READ_WRITE
+            }
+            ConnectionType::Sync => common_flags | OpenFlags::SQLITE_OPEN_READ_WRITE,
+        }
+    }
+}
+
+// XXX - move this struct to ./api.rs?
+/// The entry-point to the places API. This object gives access to database
+/// connections and other helpers. It enforces that only 1 write connection
+/// can exist to the database at once.
+pub struct PlacesAPI {
+    db_name: PathBuf,
+    encryption_key: Option<String>,
+    write_connection: Mutex<Cell<Option<PlacesDb>>>,
```

Comment:
`Mutex<Cell<...>>` is redundant. You can do what you need with just `Mutex<Option<T>>` -- acquiring a mutex gives you access to the data it guards mutably.

Reply:
Or can just use `Cell<Option<T>>` (without the Mutex), since we'll need external synchronization anyway.

Probably via another ConcurrentHandleMap, unless we want to say 'you can only have one PlacesApi', in which case we could just use a singleton `Mutex<Option<PlacesApi>>` (Or maybe `RwLock`).